File: web/lib/api/ai.ts
import apiClient from "./client";

// ─── 模型类型 ────────────────────────────────────────────────────
// "standard" = nano banane  (gemini-2.5-flash-image-preview-0924)
// "pro"      = nano banane pro (gemini-3-pro-image-preview)
export type ImageModelTier = "standard" | "pro";

// ─── 文生图 ─────────────────────────────────────────────────────
export interface Text2ImgRequest {
  prompt: string;
  /** "standard"（nano banane）| "pro"（nano banane pro） */
  model?: ImageModelTier | string;
  aspect_ratio?: "1:1" | "3:4" | "4:3" | "9:16" | "16:9";
  number_of_images?: number;
  image_size?: "1K" | "2K";
}

export interface Text2ImgResponse {
  task_id: string;
  image_url: string;
  model: string;
  enhanced_prompt?: string;
}

// ─── 图生图 / 图片编辑 ───────────────────────────────────────────
export interface ImgEditRequest {
  prompt: string;
  image: File;
}

// ─── 换脸 ───────────────────────────────────────────────────────
export interface FaceSwapRequest {
  source_image: File;
  target_image: File;
}

// ─── 文生视频 ───────────────────────────────────────────────────
export interface Text2VideoRequest {
  prompt: string;
  model?: string;
  duration?: number;
  aspect_ratio?: "16:9" | "9:16" | "1:1";
}

export interface TaskResponse {
  task_id: string;
  status: "pending" | "processing" | "completed" | "failed";
  result_url?: string;
  error?: string;
}

// ─── AI API 封装 ─────────────────────────────────────────────────
export const aiApi = {
  /** 文生图（同步，直接返回图片 URL） */
  text2img(data: Text2ImgRequest): Promise<Text2ImgResponse> {
    return apiClient.post<Text2ImgResponse>("/api/ai/text2img", data);
  },

  /** 图片编辑（multipart） */
  editImage(data: ImgEditRequest): Promise<Text2ImgResponse> {
    const form = new FormData();
    form.append("prompt", data.prompt);
    form.append("image", data.image);
    return apiClient.post<Text2ImgResponse>("/api/ai/img2img", form);
  },

  /** 换脸（multipart） */
  faceSwap(data: FaceSwapRequest): Promise<Text2ImgResponse> {
    const form = new FormData();
    form.append("source_image", data.source_image);
    form.append("target_image", data.target_image);
    return apiClient.post<Text2ImgResponse>("/api/ai/faceswap", form);
  },

  /** 文生视频（异步，返回 task_id，需轮询状态） */
  text2video(data: Text2VideoRequest): Promise<{ task_id: string }> {
    return apiClient.post<{ task_id: string }>("/api/ai/text2video", data);
  },

  /** 查询异步任务状态 */
  getTask(taskId: string): Promise<TaskResponse> {
    return apiClient.get<TaskResponse>(`/api/ai/task/${taskId}`);
  },
};

File: server/internal/handler/ai.go
package handler

import (
	"io"
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/growpilot/server/internal/middleware"
	"github.com/growpilot/server/internal/service"
)

type AIHandler struct {
	svc *service.AIService
}

func NewAIHandler(svc *service.AIService) *AIHandler {
	return &AIHandler{svc: svc}
}

// POST /api/ai/text2img
func (h *AIHandler) Text2Img(c *gin.Context) {
	userID := middleware.CurrentUserID(c)

	var req service.Text2ImgRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		BadRequest(c, err.Error())
		return
	}

	result, err := h.svc.Text2Img(c.Request.Context(), userID, req)
	if err != nil {
		InternalError(c, err.Error())
		return
	}
	OK(c, result)
}

// POST /api/ai/img2img  (multipart)
func (h *AIHandler) EditImage(c *gin.Context) {
	userID := middleware.CurrentUserID(c)

	prompt := c.PostForm("prompt")
	if prompt == "" {
		BadRequest(c, "prompt is required")
		return
	}

	file, _, err := c.Request.FormFile("image")
	if err != nil {
		BadRequest(c, "image file is required")
		return
	}
	defer file.Close()

	imageData, err := io.ReadAll(file)
	if err != nil {
		InternalError(c, "failed to read image")
		return
	}

	result, err := h.svc.EditImage(c.Request.Context(), userID, prompt, imageData, "image/jpeg")
	if err != nil {
		InternalError(c, err.Error())
		return
	}
	OK(c, result)
}

// POST /api/ai/faceswap  (multipart)
func (h *AIHandler) FaceSwap(c *gin.Context) {
	userID := middleware.CurrentUserID(c)

	sourceFile, _, err := c.Request.FormFile("source_image")
	if err != nil {
		BadRequest(c, "source_image is required")
		return
	}
	defer sourceFile.Close()

	targetFile, _, err := c.Request.FormFile("target_image")
	if err != nil {
		BadRequest(c, "target_image is required")
		return
	}
	defer targetFile.Close()

	sourceData, _ := io.ReadAll(sourceFile)
	targetData, _ := io.ReadAll(targetFile)

	// 暂时返回占位响应，待接入真实换脸服务
	_ = userID
	_ = sourceData
	_ = targetData

	OK(c, gin.H{
		"task_id":   "placeholder",
		"image_url": "",
		"model":     "faceswap-v1",
		"msg":       "faceswap service will be integrated in Phase 2",
	})
}

// POST /api/ai/text2video
func (h *AIHandler) Text2Video(c *gin.Context) {
	userID := middleware.CurrentUserID(c)

	var req service.Text2VideoRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		BadRequest(c, err.Error())
		return
	}

	taskID, err := h.svc.Text2Video(c.Request.Context(), userID, req)
	if err != nil {
		InternalError(c, err.Error())
		return
	}
	OK(c, gin.H{"task_id": taskID})
}

// GET /api/ai/task/:id
func (h *AIHandler) GetTask(c *gin.Context) {
	userID := middleware.CurrentUserID(c)
	taskID := c.Param("id")

	result, err := h.svc.GetTask(c.Request.Context(), userID, taskID)
	if err != nil {
		NotFound(c, "task not found")
		return
	}
	OK(c, result)
}

// POST /api/assistant/chat  (流式)
func (h *AIHandler) AssistantChat(c *gin.Context) {
	// Phase 4 实现：此处先返回占位响应
	c.JSON(http.StatusOK, gin.H{
		"code": 0,
		"msg":  "ok",
		"data": gin.H{"content": "AI营销助手即将上线，敬请期待！"},
	})
}
